File: DartWebClient/lib/model/map_point.dart
part of model;

class MapPoint extends LatLngWrapper {

  MapPoint(num lat, num lng) : super(lat, lng);

  MapPoint.fromMap(Map mapPointMap) : super(mapPointMap["latitude"], mapPointMap["longitude"]);

  factory MapPoint.instanceFromMap(Map mapPointMap) {
    MapPoint mapPoint;
    switch (mapPointMap["@type"]) {
      case "MapPoint":
        mapPoint = new MapPoint.fromMap(mapPointMap);
        break;
      default:
        mapPoint = new Station.instanceFromMap(mapPointMap);
    }
    return mapPoint;
  }
  
  bool isStation() => this is Station; 

  num distanceTo(MapPoint mapPoint) {
    final num radius = 6371;
    num dLatitude = toRadians(mapPoint.lat - lat);
    num dLongitude = toRadians(mapPoint.lng - lng);
    num latitude1 = toRadians(lat);
    num latitude2 = toRadians(mapPoint.lat);
    num haversine = sin(dLatitude / 2) * sin(dLatitude / 2) + sin(dLongitude / 2) * sin(dLongitude / 2) * cos(latitude1) * cos(latitude2);
    num c = 2 * atan2(sqrt(haversine), sqrt(1 - haversine));
    return radius * c * 1000;
  }

  num toRadians(num deg) => deg * (PI / 180);

  bool equals(MapPoint mapPoint) {
    return lat == mapPoint.lat && lng == mapPoint.lng;
  }

  Map toJson() {
    Map json = new Map();
    json["@type"] = runtimeType.toString();
    json["latitude"] = lat.toString();
    json["longitude"] = lng.toString();
    return json;
  }

}

File: Transportation-model/src/main/java/com/transportation/transportation/model/dtos/MapPoint.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.transportation.transportation.model.dtos;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import com.transportation.transportation.model.entities.Station;
import java.io.Serializable;
import java.util.Objects;
import javax.validation.constraints.NotNull;

/**
 *
 * @author youssef
 */
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME)
@JsonSubTypes({
    @JsonSubTypes.Type(Station.class)
})
public class MapPoint implements Serializable {

    private static final long serialVersionUID = 1L;
    @NotNull
    protected Float latitude;
    @NotNull
    protected Float longitude;

    public MapPoint() {
    }

    public MapPoint(Float latitude, Float longitude) {
        this.latitude = latitude;
        this.longitude = longitude;
    }

    public Float getLatitude() {
        return latitude;
    }

    public void setLatitude(Float latitude) {
        this.latitude = latitude;
    }

    public Float getLongitude() {
        return longitude;
    }

    public void setLongitude(Float longitude) {
        this.longitude = longitude;
    }

    /**
     * 
     * @param mapPoint
     * @return Distance in Meter
     */
    public Double distanceTo(MapPoint mapPoint) {
        final Float radius = 6371F;
        Double dLatitude = Math.toRadians(mapPoint.getLatitude() - latitude);
        Double dLongitude = Math.toRadians(mapPoint.getLongitude() - longitude);
        Double latitude1 = Math.toRadians(latitude);
        Double latitude2 = Math.toRadians(mapPoint.getLatitude());
        Double haversine = Math.sin(dLatitude / 2) * Math.sin(dLatitude / 2)
                + Math.sin(dLongitude / 2) * Math.sin(dLongitude / 2)
                * Math.cos(latitude1) * Math.cos(latitude2);
        Double c = 2 * Math.atan2(Math.sqrt(haversine), Math.sqrt(1 - haversine));
        return radius * c * 1000;
    }

    public boolean isNear(MapPoint mapPoint, Double distanceInMeter) {
        return distanceTo(mapPoint) <= distanceInMeter;
    }

    @JsonIgnore
    public boolean isStation() {
        return this instanceof Station;
    }

    @Override
    public int hashCode() {
        int hash = 5;
        hash = 97 * hash + Objects.hashCode(this.latitude);
        hash = 97 * hash + Objects.hashCode(this.longitude);
        return hash;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == null) {
            return false;
        }
        if (getClass() != obj.getClass()) {
            return false;
        }
        final MapPoint other = (MapPoint) obj;
        if (!Objects.equals(this.latitude, other.latitude)) {
            return false;
        }
        return Objects.equals(this.longitude, other.longitude);
    }
    
    @Override
    public String toString() {
        return "MapPoint{" + "latitude=" + latitude + ", longitude=" + longitude + '}';
    }

}
